lib/utils.js

```diff
@@ -605,6 +607,111 @@ function decorateWithReadConcern(command, coll, options) {
   }
 }
 
+/*
+function parseMsg(fName, msg, param = undefined) {
+  if (!msg) {
+    return;
+  }
+  let parsed = '';
+  let fillingPlaceholder = false;
+  for (let i = 0; i < msg.length; i++) {
+    const char = msg[i];
+    if (fillingPlaceholder) {
+      fillingPlaceholder = false;
+      switch (char) {
+        case 'f':
+          parsed += fName;
+          break;
+        case 'p':
+          parsed += param;
+          break;
+        default:
+          break;
+      }
+    } else {
+      if (char === '%') {
+        fillingPlaceholder = true;
+      } else {
+        parsed += char;
+      }
+    }
+  }
+  console.log(parsed);
+  return parsed;
+}
+*/
+
+function emitDeprecationWarning(msg) {
+  if (process.emitWarning) {
+    process.emitWarning(msg, 'DeprecationWarning');
+  } else {
+    console.warn(msg);
+  }
+}
+
+function messageHandler(fName, msg, param) {
+  if (msg) {
+    // return parse with message
+    // just for now, so things don't break
+    return msg;
+  }
+  const main_message = ' is deprecated and will be removed in a later version.';
+  if (typeof param === 'string') {
+    return fName + ' parameter [' + param + ']' + main_message;
+  }
+  return fName + main_message;
+}
+
+/**
+ * Deprecates a given function's parameters, the function itself, or both. Wraps the given function.
+ * Warns only once.
+ *
+ * @param {function} [fn] the target function of deprecation
+ * @param {string} [fName] function name
+ * @param {object} [config] optional configuration for warning messages and/or specifying parameters to deprecate
+ * @param {Set} [config.deprecatedParams] parameters to deprecate
+ * @param {number} [config.optionsIndex] index of options object in function arguments array
+ * @param {string} [config.msg] custom warn message to override default
+ * @param {boolean} [config.both] true if deprecating both function and parameters
+ * @return {function} modified function that warns once per deprecation, and executes original function
+ */
+function deprecate(fn, fName, config) {
+  if (process.noDeprecation === true) {
```

Comment:
Lets refactor this to be two sub-functions that work via composition. That way we can figure out the logic of the `derpecateParams` and `both` flags.